SSR The config module

samouri opened the pull request:
Calypso currently has different configuration for its different stages and the way that we handle it causes some inefficiencies.  We have _almost_ the same exact modules for wpcalypso, horizon, staging, and production.  The only difference between them is the `config` module.  This single module difference essentially forces us to separate out bundles for each environment and call the bundler 4 times within a docker build. 

This PR seeks to move the `config` module to be sent by the server as part of the initial payload which:
1. cuts build times in a down because we don't need to rebuild docker image 4 times (and have 4 different versions of the bundler: `bundle-wpcalypso.js`, `bundle-production.js`, etc.
2. No longer need to generate a config before running unit tests


Speed tests on my machine:
- Building the docker image took 4:44.62 in this branch `/try/framework/ssr-config` 
- 11:44.25 on `master`

### To Test
- [ ] start calypso using make run
- [ ] run all the unit tests (and also circle ci)
- [ ] Open https://calypso.live/?branch=try/framework/ssr-config and make sure it still works
- [ ] build the docker image.  then run it in the 4 different environments and make sure it works as expected

related issues this should solve as well: 
- Generate config at runtime:  https://github.com/Automattic/wp-calypso/pull/10608
- Always use the right config during tests: https://github.com/Automattic/wp-calypso/issues/5856

samouri commented on client/config/index.js:
```diff
@@ -0,0 +1,11 @@
+let config;
+
```
todo: add comment or readme explaining whats going on

gziolo commented:
Cool idea! Thanks for working on it. It would be great to improve config handling, in particular in the context of unit tests.

I'll double check this proposal on Monday. There might be security concern with this approach, because at least on desktop there are secret keys loaded in the server build. This code would probably expose those key to the client. It needs to be verified.

dmsnell commented on client/config/index.js:
```diff
@@ -0,0 +1,11 @@
+	config.isEnabled = window.isEnabled;
+} else {
+	throw new Error( 'config not available' );
+}
```
we can flip this conditional and then use a constant value…

```js
/**
 * Manages config flags for various deployment builds
 * @module config/index
 */

if ( 'undefined' === typeof window ||  ! window.config ) {
	throw new ReferenceError( `No configuration was found: please see client/config/README.md for more information` );
}

/**
 * @typedef {Object} CalypsoConfig
 * @property {Function} anyEnabled whether one of a set of properties is enabled
 * @property {Function} isEnabled whether a specific property is enabled
 */
export default config = {
	...window.config,
	anyEnabled: window.anyEnabled,
	isEnabled: window.isEnabled,
};
```

samouri replied:
Thanks for the docs :).  I'm definitely going to put those in. Also TIL about ReferenceError!

> Another thing to think about that may or may not come into play here is that Object.assign() sets values while destructuring assignment defines them.

`window.config` is a function here.  I don't think spread works for those...does it?  I usually prefer spread/object literal when possible.

samouri replied:
> window.config is a function here. I don't think spread works for those

Just tested it out in the online babel repl

```jsx
a = () => 5;
b = { ...a }

console.log( a() );
console.log( b() );
```

transpiled to:
```jsx
"use strict";

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

a = function a() {
  return 5;
};
b = _extends({}, a);

console.log(a());
console.log(b());
```

which printed out:
```
Uncaught TypeError: b is not a function
    at <anonymous>:1:1
```

dmsnell replied:
absolutely - I didn't realize/think about that it was a function. that's too bad, but thanks for verifying that and adjusting accordingly

alisterscott commented:
I spun this up locally in docker and ran some e2e tests against it, couldn't find any issues.
Noticed it was faster than usually spinning up in Docker, well done.

ehg commented:
One problem with this that comes to mind is that we won't have the correct build time config for each environment. For example, when building the sections we use `config` all over the place: https://github.com/Automattic/wp-calypso/blob/master/client/wordpress-com.js#L160

samouri commented:
i think @ehg is right, I can't access devdocs in the docker instance for some reason even though `config.isEnabled('devdocs') --> true`

samouri commented:
A solution i came up with that I think may be pretty valid is:
> Always make all the bundles.  

The client can request which bundles it needs.  There is one big constraint that this idea would create though:
- We cannot expect/make a bundle contain different things for different environments.  I think this is a reasonable.  If you want code to be different/only run in certain envs, it needs to be in a separate bundle

dmsnell commented:
> The client can request which bundles it needs. There is one big constraint that this idea would create though:
> 
> We cannot expect/make a bundle contain different things for different environments. I think this is reasonable. If you want code to be different/only run in certain envs, it needs to be in its own bundle

related is existing work on async loaders and changing the way we request bundles. _if_ we had a more flexible and robust system for loading code asynchronously this would be less of a concern I would imagine since we wouldn't be tied to the sections.

by the way, the section bundles have been serving us incredibly well and props to those who made them work the way they do!

gziolo commented on Dockerfile:
```diff
@@ -39,11 +39,7 @@ RUN     npm install --production || npm install --production
 COPY     . /calypso
 
 # Build javascript bundles for each environment and change ownership
```
The comment above becomes no longer valid.

samouri replied:
fixed.

gziolo commented on Dockerfile:
```diff
@@ -39,11 +39,7 @@ RUN     npm install --production || npm install --production
-          CALYPSO_ENV=stage make build-stage && \
-          CALYPSO_ENV=production make build-production && \
-          chown -R nobody /calypso
+RUN	CALYPSO_ENV=production make build && chown -R nobody /calypso
```
This need to be further investigated, because we have 3 ways of detecting env in client and server code:
* ```config( 'env )```
* ```NODE_ENV```
* ```CALYPSO_ENV```

gziolo commented on client/sections.js:
```diff
@@ -22,21 +22,20 @@ const extensionSections = extensions.map( extension => {
-		secondary: true,
-		enableLoggedOut: true
-	} );
+sections.push( {
```
I agree that we need to remove those feature flag checks from files where sections definition are located. At the same time I think we should move existing logic to entry point of each individual section. This means I'm suggesting to bring back those file guard to the first file which isn't generated by Webpack, to make sure we don't change any existing logic. We can remove some flags in separate PRs.

samouri replied:
i think both are required in this PR:
1. removing feature flag checks for build
2. adding feature flag checks to controller files

samouri replied:
@mtias yes, the bundle will be generated for any envs build. Or more correctly, this should ideally be bringing us to a single build

mtias replied:
I'm not sure generating these bundles if they are disabled at the config level is a good path. It makes extending Calypso and using it to build things harder.

gziolo replied:
How does it differ from using feature flags nested deeper into a chunk? I think it is a very good simplification that bring lots of benefits simplifying build process and drastically reducing its time.

However we need to sync it with extensions. I double checked it and it turns out that our code depends on `env_id` (https://github.com/Automattic/wp-calypso/blob/master/client/sections.js#L17). We plan to use `production` env for all non-development envs when preparing build targeted for Docker, which means it won't work as expected for `staging`, `wpcalypso` and `horizon`.

samouri replied:
@mtias: I'm not sure how generating /not-generating bundles can affect extensibility.
Whether or not the client requests the bundle is up to it.  For example, theres no reason the `devdocs` bundle can't exist on the server for `NODE_ENV=production`, as long as the client doesn't register the routes via `page`

samouri replied:
@gziolo I'll fix that now.  My current solution will be to move that check to the individual index.js and hopefully eventually we can autowrap the extensions's default exports with an `env_id` check

gziolo replied:
I think I found solution with https://github.com/Automattic/wp-calypso/pull/11352/commits/3d98ba8700a7764e54ded8fc9cfe3d0f3be14341 to this issue. This allows to bundle all sections and extensions, but at the same time never load chunks that are hidden behind the feature flag using `env_id` in `package.json` file.

@mtias @mcsf @ehg Let us know if that solution would work :)

gziolo commented:
Just wanted to summarize what we need to be aware of when working on this PR.

Calypso code depends on the following constructs:
* [NODE_ENV](https://github.com/Automattic/wp-calypso/search?utf8=%E2%9C%93&q=NODE_ENV)
* [CALYPSO_ENV](https://github.com/Automattic/wp-calypso/search?utf8=%E2%9C%93&q=CALYPSO_ENV)
* [config( 'env' ) or config( 'env_id' )](https://github.com/Automattic/wp-calypso/search?utf8=%E2%9C%93&q=%22config%28+%27env%27+%29%22&type=Code)

gziolo commented:
One more important note. We need to make sure we don't break Desktop app introducing those changes. I don't remember how config was confused in Automattic/wp-desktop but I have a feeling that it could have it's own flavors :)

Tug commented on server/config/index.js:
```diff
@@ -32,6 +34,19 @@ function anyEnabled() {
+	'<%= anyEnabled %>'
+) ( {
+	data: JSON.stringify( clientData ),
+	config: config.toString(),
```
I wouldn't worry about duplicating the config module on the client if I were you, I think we should avoid serializing functions here and depending on several globals.

1 global object containing only the raw configuration should be enough for our needs:
```
window.configData = <%= data %>;
```

Moreover, we could easily share the config module between the client and the server.
```
function loadConfig( configData = null ) {
	data = configData || window && window.configData;
}
```

gziolo replied:
Agree on this one. At the moment we have this code inlined in HTML file:

```js
function config(key) {
		if (key in data) {
			return data[key];
		}
		throw new Error('config key `' + key + '` does not exist');
	}function isEnabled(feature) {
		return !!data.features[feature];
	}function anyEnabled() {
		var args = Array.prototype.slice.call(arguments);
		return args.some(function (feature) {
			return isEnabled(feature);
		});
	}
```

samouri replied:
I just tested this out locally.  While I'd rather rely on the same module in both client/server, I couldn't figure out how to ensure webpack would exclude the serverConfig from the bundle (without changing what this file exports)

samouri replied:
I branched off a PR to test this out: https://github.com/Automattic/wp-calypso/pull/11454

gziolo replied:
I think I managed to refactor it to use shared config API. See https://github.com/Automattic/wp-calypso/pull/11352/commits/0fd7456166b81e56915f61396b4570b4b79a5983.

Tug commented on server/bundler/bin/bundler.js:
```diff
@@ -114,7 +114,7 @@ webpack( webpackConfig, function( error, stats ) {
 	} );
-	files.push( path.join( process.cwd(), 'public', 'vendor.js' ) );
-
+	files.unshift( path.join( process.cwd(), 'public', 'vendor.js' ) );
```
We usually rebase to avoid those merge commits ;)
```
git rebase origin/master
```

samouri replied:
this may not have been the best place to test out the resolve conflicts feature of GitHub.  Is there anything different (or worse) about this method?

dmsnell replied:
one at the front, one at the end…

Tug commented:
Looks good so far, good job on this @samouri :)
I had a minor comment to slightly improve the config module. Also added a few missing feature checks on the client.

I think this might need extensive testing before we are confident about a merge but I'm optimist, don't hesitate to keep pinging people to have a look!

gziolo commented on client/wordpress-com.js:
```diff
@@ -314,54 +314,44 @@ if ( config.isEnabled( 'reader' ) ) {
-		secondary: true
-	} );
-}
+sections.push( {
```
@scottsweb It looks like you created scaffolding for VIP section. It doesn't look like it is actively developed. Do you use those pages or would it be good idea to remove those files?

scottsweb replied:
We are not currently using those files and any further development there is at least 3 months out. If you would like to remove them then please do. I imagine a great deal has changed in Calypso since they were put together :)

gziolo replied:
Thanks for letting me know. I'll open a follow up PR once this one gets merged ;)

gziolo commented on client/sections.js:
```diff
@@ -22,21 +22,20 @@ const extensionSections = extensions.map( extension => {
 	}
 } );
 
-if ( config.isEnabled( 'devdocs' ) ) {
```
Check already exists in https://github.com/Automattic/wp-calypso/blob/master/client/devdocs/index.js#L13.

gziolo commented on client/wordpress-com.js:
```diff
@@ -314,54 +314,44 @@ if ( config.isEnabled( 'reader' ) ) {
+	enableLoggedOut: true
+} );
 
-if ( config.isEnabled( 'manage/custom-post-types' ) ) {
```
Check already present in https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/types/index.js#L9.

samouri commented:
Feature flags to verify/add to controllers:
- [x] devdocs
- [ ]

gziolo commented on client/wordpress-com.js:
```diff
@@ -314,54 +314,44 @@ if ( config.isEnabled( 'reader' ) ) {
 }
 
-if ( config.isEnabled( 'vip' ) ) {
-	sections.push( {
```
All VIP pages are behind features flags. See https://github.com/Automattic/wp-calypso/blob/master/client/vip/index.js.

gziolo commented:
I rebased with master to get rid of merge with master commit ;)

samouri commented:
@gziolo 
> We also need to keep in mind that we are using here Webpack DefinePlugin which is able to replace all code occurrences like process.env.NODE_ENV !== 'production' into boolean value. 

Thanks for that info! I didn't realize we were doing that.  From a `git grep 'NODE_ENV'` it looks liked we are using `process.env.NODE_ENV` 10 places within the client.  all of them reference production _or_ development which are the only two that will work going forward (so they are technically fine as-is).  The only way to reliably get the correct environment from now on will be to use `config( 'env' )`.

Do you think I should convert all usages of `NODE_ENV`/`CALYPSO_ENV` --> `config( 'env' )` and have that be the recommended way going forward?

Also are there any cases where NODE_ENV and CALYPSO_ENV are different?

samouri commented:
I don't think we actually use `config.anyEnabled` anywhere..... probably worth removing

gziolo commented on server/config/index.js:
```diff
@@ -16,11 +11,7 @@ const { serverData: data, clientData } = parser( configPath, {
-)( {
-	data: JSON.stringify( clientData )
-} );
+const ssrConfig = `var configData = ${ JSON.stringify( clientData ) };'`;
```
Yep, nice ;)

gziolo commented:
> Do you think I should convert all usages of NODE_ENV/CALYPSO_ENV --> config( 'env' ) and have that be the recommended way going forward?

I need to think about it a bit more. It would be good at least to come up with some guidelines for the future ;) What I think makes sense to keep is
```js
if ( process.env.NODE_ENV === 'development' ) { ... }
```
which is completely removed for production env. If we would refactor it to use `config( 'env' )` we would increase the size of production bundles. 

@blowery any thoughts on this?

samouri commented:
> which is completely removed for production env

I've read that in our docs in a few places but I've been struggling to figure out how this actually happens.  Do you have anything you could point me to?

gziolo commented:
> I've read that in our docs in a few places but I've been struggling to figure out how this actually happens. Do you have anything you could point me to?

It's from my experience :) Since I added this check in here https://github.com/Automattic/wp-calypso/blob/master/client/boot/index.js#L123 `devmodules` chunk is generated only for `development` env. This chunk is never created when you force production build.

blowery commented:
> Also are there any cases where NODE_ENV and CALYPSO_ENV are different?

Yup. NODE_ENV is either `production` or `development`. The latter value doesn't really matter, as we only really care about when NODE_ENV is `production`. This is used by us, but also by other libs to drop code from a "production" build. It should be unrelated to the CALYPSO_ENV.

CALYPSO_ENV is one of `development`, `wpcalypso`, `horizon`, `production`, `staging`. The overlap on the `production` key is a bit confusing, but it is what it is. 

We use CALYPSO_ENV to determine what to include in a bundle based on the deployment target.

We use NODE_ENV mostly for dead code elimination across "production" builds (which are currently all CALYPSO_ENVs !== development).

blowery commented:
Also, we don't expose CALYPSO_ENV to code being bundled, at least not in the same way we do NODE_ENV. That's handled by the DefinePlugin here: https://github.com/Automattic/wp-calypso/blob/master/webpack.config.js#L90

samouri commented on config/README.md:
```diff
@@ -15,7 +15,7 @@ console.log( config( 'redirect_uri' ) );
 -------------
 
-The config files contain a features object that can be used to determine whether to enable a feature for certain environments. This allows us to merge in-progress features without launching them to production. The config module adds two methods to detect this: `config.isEnabled()` and `config.anyEnabled()`. Please make sure to add new feature flags alphabetically so they are easy to find.
+The config files contain a features object that can be used to determine whether to enable a feature for certain environments. This allows us to merge in-progress features without launching them to production. The config module adds a methods to detect this: `config.isEnabled()`. Please make sure to add new feature flags alphabetically so they are easy to find.
```
methods --> method

samouri commented on config/README.md:
````diff
@@ -29,7 +29,5 @@ The config files contain a features object that can be used to determine whether
 
 ```bash
-ENABLE_FEATURES=manage/plugins/compatibility-warning DISABLE_FEATURES=code-splitting,reader make run
+ENABLE_FEATURES=manage/plugins/compatibility-warning DISABLE_FEATURES=code-splitting,reader
````
I think that the removal of `make run was an accident. I'll add it back in

samouri commented:
@withinboredom:
> It may be a worthy experiment to send the config function as a "hardcoded" global from the server.

I'm not sure what you mean by that.  Do you mean the data that that `config` module depends on serializing the actual module?    Initially in the beginning of the PR I had been sending the serialized function as well as data, but thanks to @Tug's suggestion and @gziolo's commit https://github.com/Automattic/wp-calypso/pull/11352/commits/a80d2dc991f3262faaf29483145f8bc556ca3068  we moved to sharing config code and only sending over the underlying data.

> Another possible experiment is using the minifier, instead of webpack, to create various bundles per environment. Do one webpack build, and have the minifier replace config() calls with the actual values, creating a separate file per environment. This could be much faster than a full build per bundle and remove the cost of the function call. The minifier might even remove the if ("production" === "development") blocks, I don't know. It might be worth investigating.

I'm also not sure what you mean here in terms of 'could be much faster than a full build per bundle'.  This PR proposes to do a single build with all the bundles w/ no variation between the different environments except for NODE_ENV=development vs. NODE_ENV=production.  

I do agree though that figuring out how to hardcode all values of config would be pretty cool and but that would also mean having different builds per environment which then reverts the whole benefit of this PR.

samouri commented:
Summary of some of the stats changing:

1. Initial Payload (the index.html sent): 6.73Kb --> 20.79Kb. This is super cacheable so I expect users to hit 304 Not Modified most of the time (~ 181bytes).
2. Bundle Sizes: I couldn't detect a change here by running `make analyze-bundles`.  That said its not easy to notice.
3. Build time of `time make docker-build`: 8:41.29 --> 4:47.62.  It is slightly less stark since the recent commit where manglification is multithreaded.

samouri commented:
All:   I think this may be ready for merging & a final round of testing.  

Left to test:
1. desktop
2. rerun the E2E tests:  @alisterscott ?

Thoughts?

alisterscott commented:
> rerun the E2E tests: @alisterscott ?

I've re-run all the e2e tests (except those outlined in #11523 which don't work), and no issues discovered.

gziolo commented on client/extensions/hello-dolly/index.js:
```diff
@@ -56,5 +59,8 @@ const render = ( context ) => {
 
 export default function() {
-	page( '/hello-dolly/:site?', siteSelection, navigation, render );
+	const calypsoEnv = config( 'env_id' );
```
We definitely need to abstract it :)

gziolo replied:
Done with https://github.com/Automattic/wp-calypso/pull/11352/commits/3d98ba8700a7764e54ded8fc9cfe3d0f3be14341.

gziolo commented:
I added my proposal for solving the last remaining issue with the extenstions and `env_id` allowing to enable a given extension for a selected envs. See https://github.com/Automattic/wp-calypso/pull/11352/commits/3d98ba8700a7764e54ded8fc9cfe3d0f3be14341 for details.

gziolo commented on client/sections.js:
```diff
@@ -14,29 +14,27 @@ const extensions = require( 'extensions' );
-			? pkg.section
-			: null;
+
+		return Object.assign( {}, pkg.section, { envId: pkg.env_id } );
```
We probably can move `env_id` in `package.json` to `section` field.

gziolo commented on server/bundler/loader.js:
```diff
@@ -134,6 +140,10 @@ function requireTemplate( section ) {
 
 		return acc.concat( [
 			'page( ' + pathRegex + ', function( context, next ) {',
+			'	var envId = ' + JSON.stringify( section.envId ) + ';',
```
There is some code duplication here, we can refactor it once we agree that this solution makes sense :)

mcsf commented on server/pages/index.js:
```diff
@@ -349,6 +349,10 @@ module.exports = function() {
 		.forEach( section => {
+			if ( section.envId && section.envId.indexOf( config( 'env_id' ) ) === -1 ) {
+				return;
+			}
```
Semantically better to make this a `.filter` that precedes the `.forEach`.

samouri replied:
agreed! will change now

gziolo replied:
Yes, sure!
I was totally focused and making it done :)

dmsnell commented on client/lib/config/index.js:
```diff
@@ -0,0 +1,35 @@
+		if ( key in data ) {
+			return data[ key ];
+		}
+		throw new Error( 'config key `' + key + '` does not exist' );
```
We should mirror the behavior changed in #11686 to prevent accidentally crashing large parts of the Calypso runtime when individual modules crash on import.

gziolo replied:
> We should mirror the behavior changed in #11686 to prevent accidentally crashing large parts of the Calypso runtime when individual modules crash on import.

Yes, we should keep that in mind. If #11686 gets merged first we should move this logic in here.

> if that PR merges first (which I hope it does) then I think it would be fine to move the code here but create `config` outside of `createConfig` and inject `data` as a curried parameter maybe so that we can still test `config()` easily

Do you mean having:
```js
const config = data => key => { ... };

const createConfig = data => {
    const localConfig = config( data );
   localConfig.isEnabled = ...
  return localConfig;
};
```

dmsnell commented on client/sections.js:
```diff
@@ -14,29 +14,27 @@ const extensions = require( 'extensions' );
-		enableLoggedOut: true
-	} );
-}
 module.exports = sections.concat( extensionSections.filter( Boolean ) );
```
is there a need to directly mutate `sections` anyway? could we get by with an immutable update that doesn't muck around with entirely separate modules?

```js
module.exports = sections
	.concat( [ {
		name: 'devdocs',
		paths: [ '/devdocs' ],
		…
	}, {
		name: 'devdocs',
		paths: [ '/devdocs/start' ],
		…
	} ] )
	.concat( extensionSections.filter( Boolean ) );
```

gziolo replied:
Personally I would move all `devdocs` to `/extensions` folder instead :)

dmsnell commented on client/signup/config/flows.js:
```diff
@@ -280,7 +280,7 @@ function filterFlowName( flowName ) {
+function filterDestination( destination ) {
 	return destination;
 }
 
```
seems like we might be able to inline this and just say something like `filterDestination: identity` or `filterDestination: a => a` from inside the "class" definition

gziolo replied:
Right, this change is here only because because this file wasn't under linter control before :)
`identity` seems to be a good idea here.

samouri commented:
I just squashed everything and then rebased from master with @dmsnell's latest changes to config.
- I added all of the new config tests to `client/lib/config/tests` as well as merged in the changes to the config functions into `createConfig`

dmsnell commented on client/lib/config/index.js:
```diff
@@ -0,0 +1,74 @@
+				'color: black' // message
+			);
+		}
+	};
```
@samouri it would be nice to keep this separated from the `createConfig` function itself. are you opposed to creating both in the module-scope and inject in the config data to `config()`?

```js
const config = data => key => { … }

const isEnabled = data => feature => { … };

module.exports = data => ( {
	config: config( data ),
	isEnabled: isEnabled( data ),
} )
```

_side note_: do we have to use `module.exports` here or can we use `export default`?

gziolo replied:
We can move `config` and `isEnabled` out, but I'm afraid definition of `createConfig` itself needs to stay as it is. `isEnabled` is exposed to the outside world as a property of `config` function.

We probably couldn't use `export default` because it is imported inside Webpack which isn't transpiled. We better check it :)

gziolo replied:
Yes, `export default` breaks build in this place.

We can refactor to the following form:

```js
module.exports = data => {
	const config = config( data );

	config.isEnabled = isEnabled( data );

	return config;
};
```

dmsnell replied:
> definition of `createConfig` itself needs to stay as it is

oh whoops. big mistake I overlooked. thanks for your code snippet though @gziolo - it captures why by bigger goal was

gziolo commented:
@samouri It looks like you have squashed commits removing history of contributions from @Tug and myself :(

samouri commented:
@gziolo, sorry about that.  I've restored correct blame history the latest force push!

samouri commented on client/wordpress-com.js:
```diff
@@ -157,7 +150,7 @@ const sections = [
 		paths: [ '/design' ],
 		module: 'my-sites/themes',
-		enableLoggedOut: config.isEnabled( 'manage/themes/logged-out' ),
+		enableLoggedOut: true,
```
Since `manage/themes/logged-out` is true in every environment configuration I think that modifying this should be harmless.  Is this true?

cc @seear

seear replied:
Correct. We should also remove `manage/themes/logged-out` from the config files if we set this to true, in case anyone tries to turn that flag off in the future expecting it to have any effect.

samouri replied:
Darn, I missed this comment.  I'll follow up with a PR after this lands

samouri commented on client/wordpress-com.js:
```diff
@@ -198,121 +191,110 @@ const sections = [
 sections.push( {
 	name: 'post-editor',
-	paths: editorPaths,
+	paths: [ '/post', '/page', '/edit' ],
```
Since manage/custom-post-types is true in every environment configuration I think that modifying this should produce no changes

cc @aduth

aduth replied:
>Since manage/custom-post-types is true in every environment configuration I think that modifying this should produce no changes

Noticing I hadn't replied here. Yes, this should be fine, but we should probably also remove the config flag altogether if we're starting to make assumptions here that it's enabled in all environments.

gziolo commented on client/lib/config/test/index.js:
```diff
@@ -50,5 +46,37 @@ describe( 'index', () => {
+				];
+
+				envs.forEach( env => {
+					process.env.NODE_ENV = env;
```
It would be nice to revert `process.env.NODE_ENV` to value set before this test suite is executed.

gziolo replied:
By the way we use only two values for NODE_ENV: `development` and `production`.

samouri replied:
I think thats what Line 54 is doing: https://github.com/Automattic/wp-calypso/pull/11352/files/f54dcdaa74f59e8249b1af19feb875b4f2c31907..baae88d6c5bb627cbf1c06117eb4ad16feeaa51e#diff-2ed3a8b7ca52e886a50c4b8ee970509eR54

samouri replied:
yea, that whole section should be modifying `CALYPSO_ENV` not `NODE_ENV`

gziolo replied:
I missed that, sorry :)

2nd comment is still valid though.

gziolo replied:
It's correct as it is in my opinion. The idea here is take advantage of dead code elimination and never execute this check in non-development ens. @dmsnell can you confirm?

gziolo commented on client/wordpress-com.js:
```diff
@@ -198,121 +191,110 @@ const sections = [
-		group: 'sites'
-	} );
-}
+sections.push( {
```
`manage/drafts` isn't enabled on production and staging. We need to add this check in `my-sites/drafts`

samouri replied:
I looked into it, and its already there :) 
https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/drafts/index.js

gziolo replied:
Yes, I missed that ;)

gziolo commented on client/wordpress-com.js:
```diff
@@ -198,121 +191,110 @@ const sections = [
+	group: 'sites'
+} );
 
-if ( config.isEnabled( 'reader' ) ) {
```
It's safe to remove. We don't have the same check inside all `reader` modules, but it is always true.

samouri commented:
There are a couple action items for after this lands:
1. Calypso.live compatibility
2. `wp-desktop` compatibility


#### Calypso.live
- this should be easy and is just changing a single string from `build-wpcalypso.js` --> `build.js`

#### wp-desktop
There is nothing fundamental about this PR that is incompatible with `wp-desktop`.  I spun up a branch that makes it compatible and it requires relatively small changes to both both repos.  The biggest issue I ran into was when both wp-desktop requires a module that exists in both repos it currently prefers wp-calypso whereas it should probably prefer itself (`/lib/config`).
- prototype prs: https://github.com/Automattic/wp-calypso/pull/11785, https://github.com/Automattic/wp-desktop/pull/275

gziolo commented on server/config/regenerate-client.js:
```diff
@@ -1,35 +0,0 @@
-	} ),
-	obj = {};
-
-keys.forEach( function( key ) {
```
@samouri I think we need to bring it back when we expose `clientData` in `server/config/parser.js`.

aduth commented:
One subtle change from this pull request is that it's no longer possible to use `debug` in wpcalypso, horizon, or staging environment because [the Webpack configuration](https://github.com/Automattic/wp-calypso/blob/4e24b236a27f9b4a892b1cd1e639de607f2f09e5/webpack.config.js#L204-L209) will now treat all non-local environments as production.

samouri commented:
@aduth, if it is important to keep `debug` in other envs, we can modify that module replacement to instead make a wrapper around debug that will no-op if `config( 'env_id' ) === 'production'`

aduth commented:
It's not my use-case, so I can't really speak to whether it's important. Just passing along an observed change from @jmosky who'd expected to be able to use debug logging in staging.

p1490989187543272-slack-calypso

jmosky commented:
It's extremely useful to be able to see what events fire as I go through WP.com. It doesn't matter to me whether I have to go to https://calypso.live/?branch=master vs. just launching the debugging tool in staging. Also, I think this might be related: https://github.com/Automattic/wp-calypso/issues/12277 .

samouri commented:
i'll whip up a pr that attempts to fix it with a check for `env_id`
